### ISSUE TITLE
Support of multiple gpu for Training

### ISSUE THREAD

**Issue by devdips123** [AntonMu/TrainYourOwnYOLO#119]
I was wondering if the code supports to be trained on multiple GPUs. I know that during detection, I can specify multiple gpus, but I didn't see any code to support multiple GPUs during training. Could you please advice?

**Comment by devdips123** [comment]
Just an update. I tried to run the code on 2 gpus. Though it runs on both gpus, but i observed that the utilization of gpu-0 is 99% while the gpu-1 utilization is always < 1%.
Could you help me figure out what the issue is and how to fix it?

**Comment by bertelschmitt** [comment]
Are you sure it is actually using the second GPU? With nvidia-smi, you should see very high memory usage for the GPU in operation with cuda.  Also, that option is only available in Detector.py, and not in Train_YOLO.py

**Comment by bertelschmitt** [comment]
O.K., I tested detection (not training) with 2 GPUs, a 1080TI, and a 1060 I had sitting around. At the first attemt, I got an error in 

`ops.py, line 4398, in _apply_device_functions  op._set_device_from_string(device_string)
AttributeError: '_TfDeviceCaptureOp' object has no attribute '_set_device_from_string'` 

That error is all over the Internet, the recommendation seems to be to downgrade to Tensorflow 1.31.1 -  my Tensorflow was at 1.15.2 . (Apparently, the code has to be rewritten to allow for multi GPU with a higher-versioned Tensorflow.)
So I downgraded, and started Detector.py with --gpu_num 2 . The detector works fine, and according to nvidia-smi, both cards are in use.

Interestingly, the 1060 draws only 35 watts in thjat setup, the 1080 twice as much. In the name of a greener planet, I might use two 1060s for this machine. It only does detection.

**Comment by bertelschmitt** [comment]
2 vs 1 GPUs
GPU 1: 1080 TI, 11GB
GPU 2:  1060, 6GB
Incoming video stream: 25 fps rtsp, realtime, each frame run through YOLO  and written to outfile

Results  2 GPU: 10 fps
Results  1 GPU: 19 fps  (used 1080 TI)

Not what one would expect .....

**Comment by bertelschmitt** [comment]
I did a few more, and less casual tests to find out whether multiple GPUs have an impact on YOLO Detection. This software package does not offer a multi GPU option. 

Executive summary: 2 are not much better than 1.

Tests were run on my dedicated detection box, i7 6700K 4200 Mhz, 64 G RAM running Ubuntu 18.04, Python3.7.7 compiled with all optimizations.

Input video H.264-encoded live RTSP stream, 1280 x 720 at 25 fps.
No output video produced for this test.

Results:

1 card, 1080TI 11G: 20-21 fps
1 card, 1060 6G: 10-11 fps
2 card, 1080TI 11G/1060 6G: 10 - 11  fps
2 card, 1060/1060 6G: 11-12  fps

Observations: 

The CPU of the machine does not get much work. TOP shows a load average of a little over 1.
In the 2 GPU cases, it has been verified with nvidia-smi that both cards are being used. In both cases, GPU memory is shown as > 90% in use

With the like-for-like 2 1060 GPU combo, the second GPU appears to work harder, showing a 57% GPU utilization and a power draw of 97 Watts (out of 120 W possible)
The first of the 2 1060 GPUs is not working hard at all, showing a 23% GPU utilization and a power draw of 27 Watts (out of 120 W possible)

Verdict: If you want higher frame rates, get a faster GPU. Adding a second GPU of the same type had little effect on the frame rate. Combining a fast 1080TI GPU with an underpowered 1060 had disastrous results, and brought the frame rate down to single 1060 levels.  

I did not test a dual 1080TI combo as my other 1080 is busy crunching models. I do not expect a much different outcome.

With a 1080TI, near-realtime detection can be produced. To keep detection from lagging behind, the input rate can be lowered (say, to 15 fps.) If that is not possible, with a few lines of code changed in detect_video(), YOLO can be told to run detection only on every 2nd frame. With that change in place, the 1060 delivered 18-20 fps, the 1080TI did the full 25 fps without breaking a sweat.
